src/rust/engine/process_execution/src/docker.rs

```diff
@@ -0,0 +1,330 @@
+            // error (something missing from the sandbox, incorrect permissions, etc).
+            //
+            // Given that this is expected to be rare, we dump the entire process definition in the
+            // error.
+            ProcessError::Unclassified(format!("Failed to execute: {}\n\n{}", req_debug_repr, msg))
+          })
+          .await;
+
+        if self.keep_sandboxes == KeepSandboxes::Always
+          || self.keep_sandboxes == KeepSandboxes::OnFailure
+            && res.as_ref().map(|r| r.exit_code).unwrap_or(1) != 0
+        {
+          workdir.keep(&req.description);
+          setup_run_sh_script(&req.env, &req.working_directory, &req.argv, workdir.path())?;
+        }
+
+        res
+      }
+    )
+    .await
+  }
+}
+
+#[async_trait]
+impl CapturedWorkdir for CommandRunner {
+  type WorkdirToken = ();
+
+  async fn run_in_workdir<'a, 'b, 'c>(
+    &'a self,
+    workdir_path: &'b Path,
+    _workdir_token: (),
+    req: Process,
+    _exclusive_spawn: bool,
+  ) -> Result<BoxStream<'c, Result<ChildOutput, String>>, String> {
+    let env = req
+      .env
+      .iter()
+      .map(|(key, value)| format!("{}={}", key, value))
+      .collect::<Vec<_>>();
+
+    let workdir_path_as_string = workdir_path
+      .to_path_buf()
+      .into_os_string()
+      .into_string()
+      .map_err(|s| {
+        format!(
+          "Unable to convert workdir_path due to non UTF-8 characters: {:?}",
+          s
+        )
+      })?;
+
+    let working_dir = req
+      .working_directory
+      .map(|relpath| Path::new("/pants-work").join(&relpath))
+      .unwrap_or_else(|| Path::new("/pants-work").to_path_buf())
+      .into_os_string()
+      .into_string()
+      .map_err(|s| {
+        format!(
+          "Unable to convert working directory due to non UTF-8 characters: {:?}",
+          s
+        )
+      })?;
+
+    // DOCKER-TODO: Set creation options so we can set platform.
+
+    let config = bollard::container::Config {
+      env: Some(env),
+      cmd: Some(req.argv),
+      working_dir: Some(working_dir),
+      host_config: Some(bollard_stubs::models::HostConfig {
+        binds: Some(vec![format!("{}:/pants-work", workdir_path_as_string)]),
+        auto_remove: Some(self.keep_sandboxes == KeepSandboxes::Never),
+        ..bollard_stubs::models::HostConfig::default()
+      }),
+      image: Some(self.image.clone()),
+      attach_stdout: Some(true),
+      attach_stderr: Some(true),
+      ..bollard::container::Config::default()
+    };
+
+    log::debug!("creating container with config: {:?}", &config);
+
+    let container = self
+      .docker
+      .create_container::<&str, String>(None, config)
+      .await
+      .map_err(|err| format!("Failed to create Docker container: {:?}", err))?;
+
+    log::debug!("created container {}", &container.id);
+
+    // DOCKER-TODO: Consider adding a drop guard to remove the container on error? (Although
```

Comment:
Will need to do something like that for the caching approach (see `nailgun::BorrowedNailgunProcess`), but easiest to wait most likely.

Reply:
As it turns out, this will need to the drop guard even in the success case. If the container is removed automatically, then this command runner may miss being able to attach to it to gather its results.

Reply:
Will add the drop guard in a follow-up.